src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java

```diff
@@ -937,12 +1058,22 @@ private int implDoFinal(ByteBuffer outBuffer)
                     doCancel = false;
                     k += token.p11.C_EncryptFinal(session.id(),
                             outAddr, outArray, (outOfs + k), (outLen - k));
+                    if (blockMode == Mode.CTS) {
+                        convertCTSVariant(outBuffer, outArray, outOfs + k);
+                    }
                 } else {
                     // Special handling to match SunJCE provider behavior
                     if (bytesBuffered == 0 && padBufferLen == 0) {
                         return 0;
                     }
-
+                    if (blockMode == Mode.CTS) {
+                        convertCTSVariant(null, padBuffer, padBufferLen);
+                        k = token.p11.C_DecryptUpdate(session.id(),
+                                0, padBuffer, 0, padBufferLen,
+                                outAddr, outArray, outOfs, outLen);
+                        outOfs += k;
+                        outLen -= k;
+                    }
```

Comment:
Why not put this block into the else block starting at line 1098 below?

Reply:
Ok.